Some performance improvements to a variety of functions

DanielVandH opened the pull request:
Here's some changes as we've discussed I wanted to look into @Kevin-Mattheus-Moerman. Ran out of time to do much else. There's some merge conflicts that I'll get to soon. I tried to space out a lot of the work into individual commits so that you can review what I've done more easily instead of looking at the whole diff all at once. Let me know what you think.

Some main breaking changes:

- In `gridpoints_equilateral`, the keyword arguments `return_faces`, `rectangular`, and `force_equilteral` now take `Val` arguments instead of Boolean arguments, so e.g. instead of passing `return_faces=true` do `return_faces=Val(true)`. This is necessary for type stability reasons.
- In `mindist`, the keyword argument `getIndex` now takes a `Val` argument instead of `Boolean`, similar to the point above. In particular, use `getIndex = Val(false)` and `getIndex = Val(true)` instead of `getIndex = false` and `getIndex = true`, respectively.
- The individual sorting functions, namely `unique_dict_index`, `unique_dict_count`, `unique_dict_inverse`, `unique_dict_index_inverse`, `unique_dict_index_count`, and `unique_dict_index_inverse_count` have all been removed. For simplicity, these are all represented by the singular `gunique` function. The keywords for this function all now take `Val` types instead of Booleans, except for `sort_entries`, which is necessary for type stability. For example, the previous behaviour `unique_x, unique_indices, inverse_indices, count = gunique(X; return_unique = true, return_index = true, return_inverse = true, return_counts = true)` is now `unique_x, unique_indices, inverse_indices, count = gunique(X; return_unique = Val(true), return_index = Val(true), return_inverse = Val(true), return_counts = Val(true))`.
- `unique_dict` has been removed. Just use `gunique(arr; compute_index = Val(true), compute_inverse = Val(true))`.
- `occursonce` now returns a `BitVector` instead of `Vector{Bool}`.
- The demos are now all also run in the test. In the future, some reference image tests / some actual tests for values should also be included. Maybe Literate.jl is a good idea for these demos so that they also serve as documentation.

Ran out of time to change much else for now but you can see that a lot of the changes are pretty similar. Other code like `n = maximum(reduce(vcat, F))` for example should also be changed into a simple non-allocating loop eventually, too. Also concerning that 
e.g. `con_vertex_edge([LineFace(1, 2), LineFace(2, 3), LineFace(1, 3)])` returns `[[1, 3], [1, 2], [2, 3]]` instead of `[(1, 3), (1, 2), (2, 3)]` (i.e., `Vector` entries instead of `Tuple`s since the size is known statically). That, and similar functions, should 
definitely be looked at eventually. The design with `sub2ind/ind2sub` seems a bit awkward, with some special casing for the case `length(siz) == length(A)` that I had to preserve.

There's also an issue with `gunique` that I've left untouched, namely that `sort_entries` doesn't actually sort the entries, it only sorts during comparison, which is not 
consistent with the docstring. I wasn't sure what the intended behaviour should be since some of the tests relied on this behaviour. If you want the entries to also be sorted 
in the returned unique values, `sort_entries ? unique(Iterators.map(_sort, X)) : unique(X)` works.

Kevin-Mattheus-Moerman commented:
@DanielVandH thanks for the help! 

> Here's some changes as we've discussed I wanted to look into @Kevin-Mattheus-Moerman. Ran out of time to do much else. There's some merge conflicts that I'll get to soon. I tried to space out a lot of the work into individual commits so that you can review what I've done more easily instead of looking at the whole diff all at once. Let me know what you think.

Perhaps a video call would be good to check some of the changes together. 

>     * In `gridpoints_equilateral`, the keyword arguments `return_faces`, `rectangular`, and `force_equilteral` now take `Val` arguments instead of Boolean arguments, so e.g. instead of passing `return_faces=true` do `return_faces=Val(true)`. This is necessary for type stability reasons.
>     * In `mindist`, the keyword argument `getIndex` now takes a `Val` argument instead of `Boolean`, similar to the point above. In particular, use `getIndex = Val(false)` and `getIndex = Val(true)` instead of `getIndex = false` and `getIndex = true`, respectively.

I don't understand why this is needed but happy to support it if it makes it better. Perhaps you can explain it to me some time. 

>     * The individual sorting functions, namely `unique_dict_index`, `unique_dict_count`, `unique_dict_inverse`, `unique_dict_index_inverse`, `unique_dict_index_count`, and `unique_dict_index_inverse_count` have all been removed. For simplicity, these are all represented by the singular `gunique` function. 

This is fine in principle. 

>     * `unique_dict` has been removed. Just use `gunique(arr; compute_index = Val(true), compute_inverse = Val(true))`.

Okay
 
>     * `occursonce` now returns a `BitVector` instead of `Vector{Bool}`.

Okay 
 
>     * The demos are now all also run in the test. In the future, some reference image tests / some actual tests for values should also be included. Maybe Literate.jl is a good idea for these demos so that they also serve as documentation.

I am unsure if this is a good idea. Some of tests already copy the demos to a large extent, and I am not sure if anything is fundamentally missing by not running the examples. Also some of the examples are quite heavy/time consuming. So I think it may be best to not have them be part of the tests. 
 
> Other code like `n = maximum(reduce(vcat, F))` for example should also be changed into a simple non-allocating loop eventually, too. 

Okay this should be easy. 

> Also concerning that e.g. `con_vertex_edge([LineFace(1, 2), LineFace(2, 3), LineFace(1, 3)])` returns `[[1, 3], [1, 2], [2, 3]]` instead of `[(1, 3), (1, 2), (2, 3)]` (i.e., `Vector` entries instead of `Tuple`s since the size is known statically). That, and similar functions, should definitely be looked at eventually. 

This may not always be possible. The example you show works but in general some nodes may touch any number of edges so some connect to 2 other to 3, or 4, or ... so any number. For instance `con_vertex_edge([LineFace(1, 2), LineFace(2, 3), LineFace(1, 3), LineFace(1, 5)])` would return `[[1, 3, 4], [1, 2], [2, 3], [ ],[4]]`. So the lengths are 3, 2, 2, 0, and 1. I suppose it could be made to become `[(1, 3, 4), (1, 2), (2, 3), (), (4)]`, but is that better? 

> The design with `sub2ind/ind2sub` seems a bit awkward, with some special casing for the case `length(siz) == length(A)` that I had to preserve.

I'll look into that. 
 
> There's also an issue with `gunique` that I've left untouched, namely that `sort_entries` doesn't actually sort the entries, it only sorts during comparison, which is not consistent with the docstring. I wasn't sure what the intended behaviour should be since some of the tests relied on this behaviour. If you want the entries to also be sorted in the returned unique values, `sort_entries ? unique(Iterators.map(_sort, X)) : unique(X)` works.

The intended behaviour is that it sorts the entries for checking uniqueness but the output is unsorted. Let's say we have a valid face with node indices [4 2 1 3]. Lets say it is clockwise face then [1 3 4 2] is another clockwise face. An inverted anti-clockwise face would be [2 4 3 1]. And each could be seen as the same face when the indices are sorted. Sorting each entry sorts each of them to [1 2 3 4]. Hence the sorting will allow us to detect them as the same. However the sorted result [1 2 3 4] is not the result we want to keep (and may not be a valid face as flipping an edge on a general polygon can make it invalid e.g. you can get a "crossed" quadrilateral). So this sorting option/functionality needs to remain intact, I sort before checking uniqueness but output the raw entries.

DanielVandH commented:
> I don't understand why this is needed but happy to support it if it makes it better. Perhaps you can explain it to me some time.

Because the returned type from the function depended on the value provided to the keyword argument. Since it is not known statically whether a `Bool` is `true` or `false`, the function is type unstable. By using `Val(true)` or `Val(false)`, it is known statically from the provided keyword argument exactly whether you want `true` or `false`.

> I am unsure if this is a good idea. Some of tests already copy the demos to a large extent, and I am not sure if anything is fundamentally missing by not running the examples. Also some of the examples are quite heavy/time consuming. So I think it may be best to not have them be part of the tests.

They seem to run pretty quickly and I think it would be important to want to ensure all demos run without errors. Up to you obviously, happy to remove.

> This may not always be possible. The example you show works but in general some nodes may touch any number of edges so some connect to 2 other to 3, or 4, or ... so any number. For instance con_vertex_edge([LineFace(1, 2), LineFace(2, 3), LineFace(1, 3), LineFace(1, 5)]) would return [[1, 3, 4], [1, 2], [2, 3], [ ],[4]]. So the lengths are 3, 2, 2, 0, and 1. I suppose it could be made to become [(1, 3, 4), (1, 2), (2, 3), (), (4)], but is that better?

Ah okay, then yes yours is better. Variable length `Tuple`s would be even worse :).

> The intended behaviour is that it sorts the entries for checking uniqueness but the output is unsorted. Let's say we have a valid face with node indices [4 2 1 3]. Lets say it is clockwise face then [1 3 4 2] is another clockwise face. An inverted anti-clockwise face would be [2 4 3 1]. And each could be seen as the same face when the indices are sorted. Sorting each entry sorts each of them to [1 2 3 4]. Hence the sorting will allow us to detect them as the same. However the sorted result [1 2 3 4] is not the result we want to keep (and may not be a valid face as flipping an edge on a general polygon can make it invalid e.g. you can get a "crossed" quadrilateral). So this sorting option/functionality needs to remain intact, I sort before checking uniqueness but output the raw entries.

Makes sense

DanielVandH commented:
I think this is ready? There's no CI setup for the tests but it seems fine locally. Do you want me to add test CI?

Kevin-Mattheus-Moerman commented:
> I think this is ready? There's no CI setup for the tests but it seems fine locally. Do you want me to add test CI?

Sure thanks!

Kevin-Mattheus-Moerman commented:
@DanielVandH sorry I do have CI set up but it is not triggered on a pull request.
 https://github.com/COMODO-research/Comodo.jl/tree/main/.github%2Fworkflows

DanielVandH commented:
I changed it now so that test.yml, your existing workflow, runs on PRs. I also added dependabot.yml to keep the actions versions up to date, but that can be removed if you don't want that

DanielVandH commented:
Almost all done but there's a remaining issue with the `demo_dualclad.jl` demo (which would not have been found without running demos in tests most likely):

```julia-repl
Running demo: demo_dualclad.jl
Demos: Error During Test at C:\Users\djv23\.julia\dev\Comodo.jl\test\runtests.jl:7346
  Got exception outside of a @test
  LoadError: UndefVarError: `Ebs` not defined in local scope
  Suggestion: check for an assignment to a local variable that shadows a global of the same name.
  Stacktrace:
    [1] dualclad(F::Vector{TriangleFace{Int64}}, V::Vector{Point{3, Float32}}, s::Float64; connectivity::Symbol)
      @ Comodo C:\Users\djv23\.julia\dev\Comodo.jl\src\functions.jl:4378
    [2] top-level scope
      @ C:\Users\djv23\.julia\dev\Comodo.jl\examples\demo_dualclad.jl:78
    [3] include
      @ .\sysimg.jl:38 [inlined]
    [4] rundemo
      @ C:\Users\djv23\.julia\dev\Comodo.jl\test\runtests.jl:7351 [inlined]
    [5] (::var"#312#314"{var"#rundemo#313", String})(demo::String)
      @ Main C:\Users\djv23\.julia\dev\Comodo.jl\test\runtests.jl:7356
    [6] foreach(f::var"#312#314"{var"#rundemo#313", String}, itr::Vector{String})
      @ Base .\abstractarray.jl:3187
    [7] macro expansion
      @ C:\Users\djv23\.julia\dev\Comodo.jl\test\runtests.jl:7354 [inlined]
    [8] macro expansion
      @ C:\Users\djv23\.julia\juliaup\julia-1.11.3+0.x64.w64.mingw32\share\julia\stdlib\v1.11\Test\src\Test.jl:1704 [inlined]
    [9] top-level scope
      @ C:\Users\djv23\.julia\dev\Comodo.jl\test\runtests.jl:7347
   [10] include(fname::String)
      @ Main .\sysimg.jl:38
   [11] top-level scope
      @ none:6
   [12] eval
      @ .\boot.jl:430 [inlined]
   [13] exec_options(opts::Base.JLOptions)
      @ Base .\client.jl:296
   [14] _start()
      @ Base .\client.jl:531
  in expression starting at C:\Users\djv23\.julia\dev\Comodo.jl\examples\demo_dualclad.jl:17
Test Summary: | Error  Total   Time
Demos         |     1      1  29.5s
ERROR: LoadError: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
in expression starting at C:\Users\djv23\.julia\dev\Comodo.jl\test\runtests.jl:7346
ERROR: Package Comodo errored during testing
```

The issue is here

```julia
function dualclad(F::Vector{NgonFace{N, TF}},V::Vector{Point{ND,TV}},s; connectivity=:face) where N where TF<:Integer where ND where TV<:Real
    Fs,Vs = scalesimplex(F,V,s) # Scaled faces

    # All non-unique mesh edges 
    E = meshedges(F;unique_only=false) 
    E_Fs = meshedges(Fs;unique_only=false)

    if connectivity == :face 
        # Compute dict to find partner edges, unique edges as well as reverse indices for unique mapping
        E_sort = sort.(E) 
        d = Dict{eltype(E),Vector{Int}}() # Use dict to keep track of used values    
        Eu = Vector{eltype(E)}()
        indReverse = Vector{Int}(undef,length(E)) 
        j=0
        for (i,e) in enumerate(E_sort)          
            if !haskey(d, e)    
                j+=1 # Increment counter        
                d[e]= [i]  
                indReverse[i] = j  # Store inverse index    
                push!(Eu, E[i]) #Grow unique set      
            else
                push!(d[e],i)    
                indReverse[i] = indReverse[d[e][1]]      
            end
        end  

        # Check for boundary faces 
        count_E2F = count_edge_face(F,Eu,indReverse)
        indBoundary = findall(isone.(count_E2F))
        Eb = Eu[indBoundary]

        indX = zeros(Int,length(Eu))
        indX[indBoundary]=1:length(indBoundary)
        indX_E = indX[indReverse]

        if !isempty(Eb)
            Ebs,Vbs = scalesimplex(Eb,V,s)
            Ebs,Vbs,_ = remove_unused_vertices(Ebs,Vbs)
            Ebs = [e.+length(Vs) for e in Ebs]
            append!(Vs,Vbs) # Append boundary edge points 
        else 
            Ebs 
        end

        Fq = Vector{QuadFace{Int}}(undef,length(Eu))
        for (i,e) in enumerate(Eu)
            ii = d[sort(e)]
            if length(ii)==2
                Fq[i] = (E_Fs[ii[1]][2],E_Fs[ii[1]][1],E_Fs[ii[2]][2],E_Fs[ii[2]][1])
            else
                ii_b = indX_E[ii[1]]
                Fq[i] = (E_Fs[ii[1]][2],E_Fs[ii[1]][1],Ebs[ii_b][1],Ebs[ii_b][2])
            end
        end
    elseif connectivity == :edge
        Eu,indReverse = gunique(E; return_unique=Val(true), return_inverse=Val(true), sort_entries=true)

        Es,V_Es = scalesimplex(Eu,V,s)
        Es = [e.+length(Vs) for e in Es]
        append!(Vs,V_Es) # Append boundary edge points 
        
        Fq = Vector{QuadFace{Int}}(undef,length(E_Fs))
        for (i,e_i) in enumerate(E_Fs)
            j = indReverse[i]
            if E[i][1] == Eu[j][1] # A first occurrence edge, order matches
                Fq[i] = (e_i[2],e_i[1],Es[j][1],Es[j][2])
            else # A second occurrence edge, needs inversion
                Fq[i] = (e_i[2],e_i[1],Es[j][2],Es[j][1])
            end
        end
    end

    return Fs,Fq,Vs
end
```

See that, some lines after the `# Check for boundary faces` comment, `Ebs` is only defined if `!isempty(Eb)`, but then later it is possible when filling up values in `Fq` for `Ebs` to accessed if `length(ii) != 2`. What is the intended behaviour here, is it a bug?

DanielVandH commented:
I've now changed it so that the demos don't run on CI

Kevin-Mattheus-Moerman commented:
@DanielVandH 

> See that, some lines after the # Check for boundary faces comment, Ebs is only defined if !isempty(Eb), but then later it is possible when filling up values in Fq for Ebs to accessed if length(ii) != 2. What is the intended behaviour here, is it a bug?

The function is only valid when the surface contains only manifold triangles. The `David.obj` example apparently features over-connected edges, i.e. edges that are connected to more than 2 faces. So I have to "cure" that surface to fix this (or keep it as a known non-manifold test model). For the moment I've added an error in `dualclad` when this is detected and removed the David example from that demo.

Kevin-Mattheus-Moerman commented:
@DanielVandH is CI the right name for the environment?

Kevin-Mattheus-Moerman commented:
@DanielVandH I also think that the demos all rely on GLMakie and the CI environment may not be configured for it (see CairoMakie and WGLMakie). So it is good not to have the demos run on CI. In addition, there may be users that do not use GLMakie and it would be good that the tests still work for them. So perhaps we should make a `rundemos` function and let it be a manual testing step that users can run, so to take it out of testing entirely?

Kevin-Mattheus-Moerman commented:
Thanks @DanielVandH
